File: labs/26-testing-evals/python/main.py
import json
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "shared" / "config"))

from openai_client import OpenAIClient, configure_logging

TEST_CASES_PATH = Path(__file__).with_name("test_cases.json")


def load_test_cases(path: Path) -> list[dict]:
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


def rule_check(answer: str, case: dict, latency_ms: float) -> dict:
    checks = {}

    keywords = case.get("expected_keywords", [])
    missing = [kw for kw in keywords if kw.lower() not in answer.lower()]
    checks["keywords"] = {"passed": not missing, "missing": missing}

    if case.get("expect_json"):
        try:
            json.loads(answer)
            checks["json"] = {"passed": True}
        except json.JSONDecodeError:
            checks["json"] = {"passed": False, "error": "not valid JSON"}

    max_latency = case.get("max_latency_ms")
    if max_latency is not None:
        checks["latency"] = {"passed": latency_ms <= max_latency, "latency_ms": latency_ms}

    all_passed = all(c["passed"] for c in checks.values())
    return {"passed": all_passed, "checks": checks}


def llm_judge(client: OpenAIClient, answer: str, reference: str) -> dict:
    prompt = f"""Rate how well the following answer matches the reference answer.
Answer: {answer}
Reference: {reference}
Respond with JSON only: {{"score": 1-10, "reason": "..."}}"""
    response = client.chat_completion(
        messages=[{"role": "user", "content": prompt}],
        response_format={"type": "json_object"},
        max_tokens=200,
        temperature=0.0,
    )
    try:
        return json.loads(response["choices"][0]["message"]["content"])
    except (json.JSONDecodeError, KeyError) as exc:
        return {"score": 0, "reason": f"Failed to parse judge response: {exc}"}


def run_agent(client: OpenAIClient, question: str) -> str:
    response = client.chat_completion(
        messages=[{"role": "user", "content": question}],
        max_tokens=200,
        temperature=0.0,
    )
    return response["choices"][0]["message"].get("content", "")


def evaluate(client: OpenAIClient, cases: list[dict]) -> dict:
    results = []
    for case in cases:
        start = time.perf_counter()
        answer = run_agent(client, case["input"])
        latency_ms = (time.perf_counter() - start) * 1000

        rule_result = rule_check(answer, case, latency_ms)
        judge_result = llm_judge(client, answer, case["reference"])

        results.append(
            {
                "id": case["id"],
                "input": case["input"],
                "answer": answer,
                "rule_passed": rule_result["passed"],
                "rule_details": rule_result["checks"],
                "judge_score": judge_result.get("score"),
                "judge_reason": judge_result.get("reason"),
            }
        )

    total = len(results)
    rule_passed = sum(1 for r in results if r["rule_passed"])
    judge_scores = [r["judge_score"] for r in results if isinstance(r["judge_score"], (int, float))]
    avg_score = sum(judge_scores) / len(judge_scores) if judge_scores else 0

    return {
        "total": total,
        "rule_pass_rate": rule_passed / total if total else 0,
        "average_judge_score": avg_score,
        "cases": results,
    }


def main():
    configure_logging()
    client = OpenAIClient()
    cases = load_test_cases(TEST_CASES_PATH)
    report = evaluate(client, cases)
    print(json.dumps(report, indent=2, ensure_ascii=False))
    print(f"\nRule pass rate: {report['rule_pass_rate']:.0%}")
    print(f"Average judge score: {report['average_judge_score']:.1f}/10")


if __name__ == "__main__":
    main()

File: labs/26-testing-evals/go/main.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"

	openaiclient "github.com/ZYZ-Labs/agent-labs/shared/config"
)

type testCase struct {
	ID              string   `json:"id"`
	Input           string   `json:"input"`
	Reference       string   `json:"reference"`
	ExpectedKeywords []string `json:"expected_keywords"`
	ExpectJSON      bool     `json:"expect_json"`
	MaxLatencyMs    *int     `json:"max_latency_ms"`
}

func loadTestCases(path string) ([]testCase, error) {
	b, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var cases []testCase
	if err := json.Unmarshal(b, &cases); err != nil {
		return nil, err
	}
	return cases, nil
}

func ruleCheck(answer string, c testCase, latencyMs float64) map[string]interface{} {
	checks := map[string]interface{}{}

	var missing []string
	lower := strings.ToLower(answer)
	for _, kw := range c.ExpectedKeywords {
		if !strings.Contains(lower, strings.ToLower(kw)) {
			missing = append(missing, kw)
		}
	}
	checks["keywords"] = map[string]interface{}{"passed": len(missing) == 0, "missing": missing}

	if c.ExpectJSON {
		if err := json.Unmarshal([]byte(answer), new(interface{})); err != nil {
			checks["json"] = map[string]interface{}{"passed": false, "error": "not valid JSON"}
		} else {
			checks["json"] = map[string]interface{}{"passed": true}
		}
	}

	if c.MaxLatencyMs != nil {
		checks["latency"] = map[string]interface{}{"passed": latencyMs <= float64(*c.MaxLatencyMs), "latency_ms": latencyMs}
	}

	allPassed := true
	for _, v := range checks {
		if m, ok := v.(map[string]interface{}); ok {
			if passed, ok := m["passed"].(bool); ok && !passed {
				allPassed = false
			}
		}
	}
	return map[string]interface{}{"passed": allPassed, "checks": checks}
}

func llmJudge(client *openaiclient.Client, answer, reference string) map[string]interface{} {
	prompt := fmt.Sprintf(`Rate how well the following answer matches the reference answer.
Answer: %s
Reference: %s
Respond with JSON only: {"score": 1-10, "reason": "..."}`, answer, reference)
	resp, err := client.ChatCompletion(openaiclient.ChatCompletionRequest{
		Messages:       []openaiclient.Message{{Role: "user", Content: prompt}},
		ResponseFormat: map[string]string{"type": "json_object"},
		MaxTokens:      intPtr(200),
		Temperature:    floatPtr(0.0),
	})
	if err != nil {
		return map[string]interface{}{"score": 0, "reason": fmt.Sprintf("judge call failed: %v", err)}
	}
	var result map[string]interface{}
	if err := json.Unmarshal([]byte(resp.Choices[0].Message.Content), &result); err != nil {
		return map[string]interface{}{"score": 0, "reason": fmt.Sprintf("failed to parse judge response: %v", err)}
	}
	return result
}

func runAgent(client *openaiclient.Client, question string) string {
	resp, err := client.ChatCompletion(openaiclient.ChatCompletionRequest{
		Messages:    []openaiclient.Message{{Role: "user", Content: question}},
		MaxTokens:   intPtr(200),
		Temperature: floatPtr(0.0),
	})
	if err != nil {
		return fmt.Sprintf("Error: %v", err)
	}
	return resp.Choices[0].Message.Content
}

func evaluate(client *openaiclient.Client, cases []testCase) map[string]interface{} {
	var results []map[string]interface{}
	for _, c := range cases {
		start := time.Now()
		answer := runAgent(client, c.Input)
		latencyMs := float64(time.Since(start).Nanoseconds()) / 1e6

		ruleResult := ruleCheck(answer, c, latencyMs)
		judgeResult := llmJudge(client, answer, c.Reference)

		results = append(results, map[string]interface{}{
			"id":           c.ID,
			"input":        c.Input,
			"answer":       answer,
			"rule_passed":  ruleResult["passed"],
			"rule_details": ruleResult["checks"],
			"judge_score":  judgeResult["score"],
			"judge_reason": judgeResult["reason"],
		})
	}

	total := len(results)
	rulePassed := 0
	var scoreSum float64
	var scoreCount int
	for _, r := range results {
		if r["rule_passed"].(bool) {
			rulePassed++
		}
		if s, ok := r["judge_score"].(float64); ok {
			scoreSum += s
			scoreCount++
		}
	}
	avg := 0.0
	if scoreCount > 0 {
		avg = scoreSum / float64(scoreCount)
	}

	return map[string]interface{}{
		"total":               total,
		"rule_pass_rate":      float64(rulePassed) / float64(total),
		"average_judge_score": avg,
		"cases":               results,
	}
}

func floatPtr(f float64) *float64 {
	return &f
}

func intPtr(i int) *int {
	return &i
}

func main() {
	client := openaiclient.NewClient()
	cases, err := loadTestCases("../python/test_cases.json")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to load test cases: %v\n", err)
		os.Exit(1)
	}
	report := evaluate(client, cases)
	b, _ := json.MarshalIndent(report, "", "  ")
	fmt.Println(string(b))
	fmt.Printf("\nRule pass rate: %.0f%%\n", report["rule_pass_rate"].(float64)*100)
	fmt.Printf("Average judge score: %.1f/10\n", report["average_judge_score"].(float64))
}
